Update examples for 2.0

JohnsonKuan opened the pull request:
This PR is to update the repo so that selected examples work with the Cleanlab 2.0 API.

- Move "old" examples to `contrib/v1/` folder (code that only works with 1.0 API) - these are the examples to reproduce the results in the Confident Learning paper
- Update remaining examples so they work with 2.0 API
- Use sklearn GridSearchCV for the model selection hyperparameter optimization notebook
- Format the code (with Black) in the Jupyter notebooks
- Provide `requirements.txt `file and instructions to install requirements in a Python virtual environment
- Update `README.md` with complete table of contents with descriptions for each example (ordered by difficulty)
- Provide Python and Bash script to run all notebooks in order (optional for the user)

JohnsonKuan committed:
Create contrib folder and move old examples there. Update remaining examples in root to work with Cleanlab 2.0. Update README with table of contents, description for each example, and instructions. Add requirements file.

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update v1 README

JohnsonKuan committed:
Update requirements.txt

JohnsonKuan committed:
Update requirements.txt

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update notebooks and run_all_notebooks.py script

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Cleanup docstring

jwmueller commented:
First 2 sentences of [classifier_comparison.ipynb](https://github.com/cleanlab/examples/blob/5f73b8da41751760c8d1858f9e2c94e971e713df/classifier_comparison.ipynb) seem oddly formatted (seems like a header when it shouldn't be).

jwmueller commented on README.md:
````diff
@@ -1,21 +1,51 @@
+./run_all_notebooks.sh
+```
+
+Python script:
+
+```console
+python run_all_notebooks.py
+```
````
[Nit] I think can only mention the bash script and skip mention of this one in Readme (even though they're basically the same)

JohnsonKuan replied:
That sounds good. I removed the mention of the Python script.

9f1a549f924f08c474b0b8bdd1d9c62c5fa3f0f4

jwmueller commented on README.md:
````diff
@@ -1,21 +1,51 @@
+```
+
+Run the notebooks individually or run either scripts below which will execute and save each notebook.
+
````
Would add:

To run the examples with your own version of `cleanlab`, simply edit the first line of **requirements.txt**, or delete it and install `cleanlab` separately.

jwmueller commented on README.md:
```diff
@@ -1,21 +1,51 @@
+| ----------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
+| [iris_simple_example.ipynb](https://github.com/cleanlab/examples/blob/master/iris_simple_example.ipynb)                                         | Use `cleanlab` to find synthetic label errors in the Iris dataset.                                                     |
+| [classifier_comparison.ipynb](https://github.com/cleanlab/examples/blob/master/classifier_comparison.ipynb)                                     | Demonstrate how `cleanlab` can be used to train 10 different classifiers on 4 dataset distributions with label errors. |
+| [model_selection_demo.ipynb](https://github.com/cleanlab/examples/blob/master/model_selection_demo.ipynb)                                       | Perform hyperparameter optimization with `cleanlab`'s hyperparameters.                                                 |
```
[Nit] hyperparameter optimization to find the best settings of  `cleanlab`'s optional parameters.

JohnsonKuan replied:
That sounds better. I changed it to this description. Thanks.

9f1a549f924f08c474b0b8bdd1d9c62c5fa3f0f4

jwmueller commented on README.md:
```diff
@@ -1,21 +1,51 @@
+
+## Instructions
+
+To run the latest example notebooks, execute the commands below which will install the required libraries in a virtual environment.
```
Would add:

To run the examples with your own version of `cleanlab`, simply edit the first line of **requirements.txt**, or delete it and install `cleanlab` separately.

JohnsonKuan replied:
Sounds good. Added this. Thanks.

jwmueller commented on README.md:
````diff
@@ -1,21 +1,51 @@
+python run_all_notebooks.py
+```
+
+## Old Examples
````
[Nit] Old -> Older

JohnsonKuan replied:
Good catch, agree this should be "Older". I made this change. Thanks.

jwmueller commented on README.md:
```diff
@@ -1,21 +1,51 @@
+
+## Old Examples
+
+See the `contrib` folder for examples from v1 of `cleanlab`.
```
add: which may be helpful for reproducing results from the [Confident Learning paper](link)

JohnsonKuan replied:
Good call. Added.

jwmueller reviewed (approved):
Looks great!  Will let @cgnorthcutt do a final pass and merge.

JohnsonKuan committed:
Update README. Change format of header for classifier_comparison.ipynb.

JohnsonKuan committed:
Update README

JohnsonKuan commented:
> First 2 sentences of [classifier_comparison.ipynb](https://github.com/cleanlab/examples/blob/5f73b8da41751760c8d1858f9e2c94e971e713df/classifier_comparison.ipynb) seem oddly formatted (seems like a header when it shouldn't be).

Good catch. I agree. I changed the format for those 2 sentences.

JohnsonKuan committed:
Rename LearningWIthNoisyLabels to CleanLearning everywhere

JohnsonKuan committed:
Update requirements

JohnsonKuan commented:
Added commit to rename `LearningWithNoisyLabels` to `CleanLearning` everywhere following merge below.

https://github.com/cleanlab/cleanlab/pull/177

jwmueller commented:
Note this PR: https://github.com/cleanlab/cleanlab/pull/181

has moved: coteaching.py and fasttext.py into `cleanlab.experimental/`
so you'll need to update the corresponding examples accordingly.

JohnsonKuan committed:
Change cleanlab.filter.keep_at_least_n_per_class to _keep_at_least_n_per_class()

JohnsonKuan commented:
> cleanlab/cleanlab#181

Confirming that moving `coteaching.py` and `fasttext.py` to `cleanlab.experimental/` only impact the older examples in `contrib/v1` folder so no changes are needed to address this.

The only change needed is to update `keep_at_least_n_per_class` to `_keep_at_least_n_per_class`. I added latest commit to make this update. Confirming that all the 2.0 examples run successfully with this latest update.

15609a9

jwmueller commented:
Note we have PR to move rest of cleanlab.example_models -> cleanlab.experimental
so will need to update examples/ accordingly.

https://github.com/cleanlab/cleanlab/pull/184

jwmueller commented:
> Confirming that moving `coteaching.py` and `fasttext.py` to `cleanlab.experimental/` only impact the older examples in `contrib/v1` folder so no changes are needed to address this.

@JohnsonKuan We want to have at least one example that works with each of the files in: `cleanlab.experimental`.  So need to move some of this stuff out of contrib/v1 and make it work with the master cleanlab to have at least one example for each thing in `cleanlab.experimental` (there are no other tutorials/scripts for these files anywhere).  
Sorry I didn't think of this before!

JohnsonKuan committed:
Add example for cifar CNN and coteaching experimental modules

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update README

JohnsonKuan commented:
Added CIFAR-10 example to demonstrate the use of two `cleanlab` experimental modules below:

- [cleanlab.experimental.cifar_cnn.py](https://github.com/cleanlab/cleanlab/blob/master/cleanlab/experimental/cifar_cnn.py)
- [cleanlab.experimental.coteaching.py](https://github.com/cleanlab/cleanlab/blob/master/cleanlab/experimental/coteaching.py)

To simplify the process of going thru this example for the user, I created a bash script that downloads all the needed files for the example (training data, noisy labels, pre-computed `pred_probs`). I also simplified the readme.

JohnsonKuan commented:
Still a work-in-progress to add examples that work with 2.0 for below experimental modules:

- fasttext.py
- mnist_pytorch.py

JohnsonKuan committed:
Raise ValueError if epochs < num_gradual for coteaching

JohnsonKuan committed:
Add example for cleanlab.experimental.mnist_pytorch

JohnsonKuan committed:
Update README.md

JohnsonKuan commented:
Added example to demonstrate the use of `cleanlab` experimental module: mnist_pytorch.py

JohnsonKuan committed:
Add example for fasttext

JohnsonKuan committed:
Merge branch 'update-for-v2' of github.com:JohnsonKuan/examples into update-for-v2

JohnsonKuan commented:
Added example to demonstrate the use of `cleanlab` experimental module: fasttext.py

We now have an example for each of the experimental modules:

- cifar_cnn.py
- coteaching.py
- mnist_pytorch.py
- fasttext.py

JohnsonKuan committed:
Change cleanlab.noise_generation to cleanlab.benchmarking.noise_generation

JohnsonKuan committed:
Rename cleanlab.util to cleanlab.internal.util

JohnsonKuan committed:
Cleanup README

JohnsonKuan committed:
Add relative links to table in README

JohnsonKuan committed:
Update README.md

JohnsonKuan committed:
Update README.md

JohnsonKuan committed:
Update README.md

JohnsonKuan committed:
Update README.md

JohnsonKuan committed:
Update README.md

JohnsonKuan committed:
Update README.md

JohnsonKuan committed:
Change column name to "Example" in README table

JohnsonKuan committed:
Create separate cuda_requirements.txt file for examples that require GPU for training

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update all README

JohnsonKuan committed:
Update all README

JohnsonKuan commented:
@jwmueller @cgnorthcutt 

Confirming I added an example for each experimental module.

I also reran the examples just now to make sure they work with latest cleanlab master branch.

Ready for final review before merge.

jwmueller commented on README.md:
```diff
@@ -1,21 +1,52 @@
-Not sure where to start? Try checking out how to find [ImageNet Label Errors](imagenet/imagenet_train_label_errors.ipynb).
+This repo contains code examples that demonstrate how to use [cleanlab](https://github.com/cleanlab) and how [confident learning](https://arxiv.org/abs/1911.00068) works to find label errors.
 
+## Latest Examples
```
"Table of Contents" may be better header?

"Latest Examples" makes it sound like this stuff is very new when it is not.

JohnsonKuan replied:
That's true. "Table of Contents" sounds good to me. I made the change.

jwmueller commented on README.md:
```diff
@@ -1,21 +1,52 @@
+| 8   | [amazon-reviews-fasttext](amazon-reviews-fasttext)                                             | Demonstrate the use of the following experimental module from cleanlab: [fasttext.py](https://github.com/cleanlab/cleanlab/blob/master/cleanlab/experimental/fasttext.py)                                                                                                    |
+
+## Instructions
+
```
should specify which version of cleanlab is recommended here (see slack discussion).

JohnsonKuan replied:
Updated instructions for user to use the latest stable cleanlab release. Also updated the `requirements.txt` file so it just installs `cleanlab` instead of pointing to master branch.

jwmueller commented on README.md:
````diff
@@ -1,21 +1,52 @@
+Bash script:
+
+```console
+$ ./run_all_notebooks.sh
````
may be better as:
```
bash ./run_all_notebooks.sh
```

Some systems throw permissions issues when the `bash` is omitted.

JohnsonKuan replied:
Sounds good, agree. I made this change.

JohnsonKuan committed:
Update README to recommend use of latest stable cleanlab release

jwmueller commented on README.md:
```diff
@@ -1,21 +1,52 @@
+# cleanlab Examples
 
-Not sure where to start? Try checking out how to find [ImageNet Label Errors](imagenet/imagenet_train_label_errors.ipynb).
+This repo contains code examples that demonstrate how to use [cleanlab](https://github.com/cleanlab) and how [confident learning](https://arxiv.org/abs/1911.00068) works to find label errors.
```
Add sentence:

To quickly learn the basics of running cleanlab  on your own data, we recommend first starting [here](https://docs.cleanlab.ai/v2.0.0/index.html#quickstart) before diving into the examples below.

jwmueller replied:
TODO: that link I posted is intended to be for currently-unreleased  2.0.0 docs, verify it actually works once 2.0.0 is tagged

JohnsonKuan replied:
Sounds good, I added the sentence and will verify the link works once 2.0.0 is tagged.

JohnsonKuan replied:
@jwmueller I updated the link here to use the v2.0.0 docs and verified it.

685d5c1a6b35154659e807f9b356c8570388e57f

jwmueller reviewed (commented):
Looking good! I was able to run them successfully as well.

I think it's ready to merge, but will require a couple updates once 2.0.0 is tagged before we merge this PR. 
Please just ping me for one final look-over then.

JohnsonKuan committed:
Update README

JohnsonKuan committed:
Update all README

JohnsonKuan committed:
Update docs link to use v2.0.0

jwmueller reviewed (approved):
LGTM; I'll make a 1-2 minor edits, but great stuff!!

jwmueller committed:
use stabler links
